File: server/src/main/com/broll/gainea/server/core/stats/GameStatistic.java
package com.broll.gainea.server.core.stats;

import com.broll.gainea.net.NT_GameStatistic;
import com.broll.gainea.net.NT_RoundStatistic;
import com.broll.gainea.server.core.GameContainer;
import com.broll.gainea.server.core.player.Player;
import com.broll.gainea.server.core.processing.GameUpdateReceiverAdapter;
import com.broll.gainea.server.core.utils.GameUtils;

import java.util.ArrayList;
import java.util.List;

public class GameStatistic extends GameUpdateReceiverAdapter {

    private GameContainer game;
    private List<TurnStatistic> turnStatistics = new ArrayList<>();

    public GameStatistic(GameContainer game) {
        this.game = game;
    }

    @Override
    public void roundStarted() {
        registerRound();
    }

    public void registerRound() {
        this.turnStatistics.add(calcTurnStatistic());
    }

    public void sendStatistic() {
        calcTurnStatistic();
        NT_GameStatistic nt = new NT_GameStatistic();
        nt.rounds = turnStatistics.stream().map(TurnStatistic::get).toArray(NT_RoundStatistic[]::new);
        GameUtils.sendUpdate(game, nt);
    }

    private TurnStatistic calcTurnStatistic() {
        TurnStatistic statistic = new TurnStatistic();
        statistic.players = game.getAllPlayers().stream().map(this::calcPlayerStatistic).toArray(PlayerStatistic[]::new);
        return statistic;
    }

    private PlayerStatistic calcPlayerStatistic(Player player) {
        PlayerStatistic statistic = new PlayerStatistic();
        statistic.controlledLocations = (short) player.getControlledLocations().size();
        statistic.cards = (short) player.getCardHandler().getCards().size();
        statistic.stars = (short) player.getGoalHandler().getStars();
        statistic.points = (short) player.getGoalHandler().getScore();
        player.getUnits().forEach(unit -> {
            statistic.units++;
            statistic.totalPower += unit.getPower().getValue();
            statistic.totalHealth += unit.getHealth().getValue();
        });
        return statistic;
    }

}

File: server/src/main/com/broll/gainea/server/core/stats/GameStatistic.kt
package com.broll.gainea.server.core.stats

import com.broll.gainea.net.NT_GameStatistic
import com.broll.gainea.server.core.Game
import com.broll.gainea.server.core.player.Player
import com.broll.gainea.server.core.processing.GameUpdateReceiverAdapter
import com.broll.gainea.server.core.utils.sendUpdate


class GameStatistic(private val game: Game) : GameUpdateReceiverAdapter() {
    private val turnStatistics = mutableListOf<TurnStatistic>()
    override fun roundStarted() {
        registerRound()
    }

    fun registerRound() {
        turnStatistics.add(calcTurnStatistic())
    }

    fun sendStatistic() {
        calcTurnStatistic()
        val nt = NT_GameStatistic()
        nt.rounds = turnStatistics.map { it.nt() }.toTypedArray()
        game.sendUpdate(nt)
    }

    private fun calcTurnStatistic() = TurnStatistic(game.allPlayers.map { calcPlayerStatistic(it) })

    private fun calcPlayerStatistic(player: Player): PlayerStatistic {
        val statistic = PlayerStatistic()
        statistic.controlledLocations = player.controlledLocations.size.toShort()
        statistic.cards = player.cardHandler.cards.size.toShort()
        statistic.stars = player.goalHandler.stars.toShort()
        statistic.points = player.goalHandler.score.toShort()
        player.units.forEach {
            statistic.units++
            statistic.totalPower += it.power.value
            statistic.totalHealth += it.health.value
        }
        return statistic
    }

}
